Change default location for `bundle install`

viroulep opened the pull request:
Fixes #840.

Make `username` do the `bundle install` to `~/.bundle`.

Also toggle the [deployment](http://bundler.io/man/bundle-install.1.html#DEPLOYMENT-MODE) flag when installing in production.
(the man says it "optimizes" for deployment, so let's do it...)

I set the path to `/home/user/.bundle` instead of `vendor/bundle`:
- to avoid potential conflict when using the same repo for both local and vagrant development
- because for some reason it makes running `bundle exec` extremely slow (when ran from inside the VM, with gems installed on the filesystem shared with the host). `strace`-ing it shows an insane amount of `open` syscall, so I guess it's the source of the slow down, but I don't know exactly why: both @jfly and I have a ssd and for him it's not awfully slow. He probably got a Ferrari while I have a Lada.

jfly commented:
> He probably got a Ferrari while I have a Lada

<3 my x1 carbon gen3

jfly commented:
LGTM! As discussed offline, I would like to spin up a new production server when we merge this, rather than just re-running chef on our current production box. @viroulep, would you be up for trying to do that? If not, both @coder13 and I can do it.

viroulep commented:
I'd be up to do that but with a backup competent person to save me in case everything go wrong :)

and wtf travis?!

jonatanklosko commented on chef/site-cookbooks/wca/recipes/default.rb:
```diff
@@ -323,7 +323,8 @@
 
 #### Initialize rails gems/database
-execute "bundle install --without none" do
+execute "bundle install #{"--deployment" if rails_env == "production"} --path /home/#{username}/.bundle" do
```
Shouldn't it be `--without test development` as well?

viroulep replied:
You're right, it installs everything otherwise...
So basically we also want to run "--without production" when in development.

viroulep replied:
So I have no idea why but starting up a VM in "development" environment with "--without production" fails at the db:setup [here](https://github.com/cubing/worldcubeassociation.org/blob/a3948bde75de9cce3233192c8292d791399d871d/chef/site-cookbooks/wca/recipes/default.rb#L334) :(
I'm going to see how it behaves starting a production VM...

jonatanklosko commented:
@viroulep I've re-ran Travis and I'll fix the test in a PR now, this is unrelated thing.

viroulep commented:
Thanks, it was the weirdest bug :)

viroulep commented:
Updated, I couldn't get `--without production` to work in development, nor figure out why...
But the `--without test development` works fine in prod, so I just dropped the flag in dev :p
I also moved i18n-tasks to the dev group, because it's not a prod thing.

jonatanklosko commented:
Cool, LGTM!

viroulep commented:
I did one more change after testing production locally (change the bundle install in the deploy.sh).

For the record here are some tweaking I had to do to properly test everything:
- Manually change nginx and restart it (replace [here](https://github.com/cubing/worldcubeassociation.org/blob/a3948bde75de9cce3233192c8292d791399d871d/chef/site-cookbooks/wca/templates/worldcubeassociation.org.conf.erb#L21-L33) by a single `listen 2331;`, comment out the stuff [here](https://github.com/cubing/worldcubeassociation.org/blob/a3948bde75de9cce3233192c8292d791399d871d/chef/site-cookbooks/wca/templates/worldcubeassociation.org.conf.erb#L283-L301))
- Force the production environment [here](https://github.com/cubing/worldcubeassociation.org/blob/a3948bde75de9cce3233192c8292d791399d871d/scripts/deploy.sh#L27-L31), otherwise rake starts loading dev gems and fails :s (I did this to test a well-formed Gemfile+Gemfile.lock would install properly by running the deploy script)

I merged this, #747, and #805 before testing this, and everything worked quite nicely :)

jfly commented:
That's wonderful news, Philippe!

jfly commented:
Would you like to wait to merge up #747 and then we can spin up a new server with everything? I'll take a look at #747 right now.
